prio/encrypt.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2018, Henry Corrigan-Gibbs
+ * 
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. 
+ */
+
+#include <nss/keyhi.h>
+#include <nss/keythi.h>
+#include <nss/pk11pub.h>
+#include <prerror.h>
+
+#include "encrypt.h"
+#include "prio/rand.h"
+#include "prio/util.h"
+
+// Use Curve25519
+#define CURVE_OID_TAG SEC_OID_CURVE25519
+#define CURVE25519_KEY_LEN 32
+
+// Use 96-bit IV
+#define GCM_IV_LEN_BYTES 12
+// Use 128-bit auth tag
+#define GCM_TAG_LEN_BYTES 16
+
+#define PRIO_TAG "PrioPacket"
+#define AAD_LEN (strlen (PRIO_TAG) + CURVE25519_KEY_LEN + GCM_IV_LEN_BYTES)
+
+static SECStatus 
+derive_dh_secret (PK11SymKey **shared_secret, PrivateKey priv, PublicKey pub)
+{
+  if (priv == NULL) return SECFailure;
+  if (pub == NULL) return SECFailure;
+  if (shared_secret == NULL) return SECFailure;
+
+  SECStatus rv = SECSuccess;
+  *shared_secret = NULL;
+
+  P_CHECKA (*shared_secret = PK11_PubDeriveWithKDF (priv, pub, PR_FALSE,
+        NULL, NULL, CKM_ECDH1_DERIVE, CKM_AES_GCM, 
+        CKA_ENCRYPT | CKA_DECRYPT, 16,
+        CKD_SHA256_KDF, NULL, NULL));
+
+
+cleanup:
+  return rv;
+}
+
+SECStatus
+PublicKey_import (PublicKey *pk, const unsigned char *data, unsigned int dataLen)
+{
+  SECStatus rv = SECSuccess;
+  PrivateKey eph_priv = NULL;
+  *pk = NULL;
+
+  if (dataLen != CURVE25519_KEY_LEN)
+    return SECFailure;
+
+  // TODO: Horrible hack. We should be able to import a public key
```

Comment:
You can import keys with pk8 (private) and spki (public). See [1] for example.
This isn't great. But that's the official format :disappointed: 

[1] https://searchfox.org/nss/source/gtests/pk11_gtest/pk11_curve25519_unittest.cc#66

Reply:
Since this has to go through PK11 NSS can't take byte blobs unfortunately.

> For example, I could hardcode an SPKI representation of the all-zeros curve25519 public key, import that using SECKEY_DecodeDERSubjectPublicKeyInfo and then set the bytes of the key as I am doing now. Would that work?

That would be pretty ugly but should work :slightly_smiling_face:

Reply:
Done.